lib/ansible/galaxy/collection.py

```diff
@@ -1008,6 +1064,47 @@ def reset_stat(tarinfo):
         display.display('Created collection for %s at %s' % (collection_name, to_text(b_tar_path)))
 
 
+def _build_collection_dir(b_collection_path, b_collection_output, collection_manifest, file_manifest):
+    """ Build a collection directory from the manifest data. This should follow the same pattern as _build_collection_tar. """
+    os.makedirs(b_collection_output, mode=0o0755)
+
+    files_manifest_json = to_bytes(json.dumps(file_manifest, indent=True), errors='surrogate_or_strict')
+    collection_manifest['file_manifest_file']['chksum_sha256'] = secure_hash_s(files_manifest_json, hash_func=sha256)
+    collection_manifest_json = to_bytes(json.dumps(collection_manifest, indent=True), errors='surrogate_or_strict')
+
+    # Write contents to the files
+    for name, b in [('MANIFEST.json', collection_manifest_json), ('FILES.json', files_manifest_json)]:
+        b_path = os.path.join(b_collection_output, to_bytes(name, errors='surrogate_or_strict'))
+        b_io = BytesIO(b)
+        with open(b_path, 'wb') as file_obj:
+            shutil.copyfileobj(b_io, file_obj)
```

Comment:
BytesIO is also a file-like object, so use it with CM. Otherwise, you'll end up with the `ResourceWarning: unclosed file <BytesIO(...)>...` (it's not printed out w/o smth like `-Xdev` but it's there):
```suggestion
        with open(b_path, 'wb') as file_obj, BytesIO(b) as b_io:
            shutil.copyfileobj(b_io, file_obj)
```

Reply:
(I don't remember if Py 2.7 can do two CMs in one `with`, if not — change it to two nested blocks)

Reply:
Looks like it works on Py 2.7.